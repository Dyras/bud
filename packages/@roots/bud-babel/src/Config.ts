import type {Framework} from '@roots/bud-framework'
import {boundMethod as bind} from 'autobind-decorator'
import {isString} from 'lodash'

/**
 * Configure Babel transpiler plugin & presets
 *
 * @example
 * ```js
 * app.babel.setPlugins([
 *  ['@babel/plugin-transform-runtime', {helpers: false}],
 *  '@babel/plugin-proposal-object-rest-spread',
 *  '@babel/plugin-syntax-dynamic-import',
 *  '@babel/plugin-proposal-class-properties',
 * ])
 * ```
 *
 * @public @extension @config
 */
export interface Config {
  /**
   * Registered babel plugins
   *
   * @public
   */
  plugins: Registry

  /**
   * Registered babel presets
   *
   * @public
   */
  presets: Registry

  /**
   * Add a babel plugin
   *
   * @example
   * ```js
   * babel.setPlugin(MyPlugin, {plugin: 'options'})
   * ```
   *
   * @public
   */
  setPlugin(plugin: Registrable): Config

  /**
   * Add babel plugins
   *
   * @public
   */
  setPlugins(plugins: Array<Registrable>): Config

  /**
   * Set the options for a plugin
   *
   * @public
   */
  setPluginOptions(plugin: string, options: any): Config

  /**
   * Add a babel preset
   *
   * @example
   * ```js
   * babel.setPlugin(MyPlugin, {plugin: 'options'})
   * ```
   *
   * @public
   */
  setPreset(preset: Registrable): Config

  /**
   * Add babel presets
   *
   * @public
   */
  setPresets(presets: Array<NormalizedPlugin | string>): Config

  /**
   * Set the options for a preset
   *
   * @public
   */
  setPresetOptions(preset: string, options: any): Config
}

export type Options = {
  plugins?: Plugin[]
  config?: boolean | string
}

export type NormalizedPlugin = [string, {[key: string]: any}]

export type Plugin = string | NormalizedPlugin | CallableFunction

export type Registrable = string | NormalizedPlugin

export interface Registry {
  [key: string]: [string, any]
}

export class Config {
  /**
   * @public
   */
  public name = '@roots/bud-babel'

  public _app: () => Framework

  public plugins: Registry = {}

  public presets: Registry = {}

  public get app() {
    return this._app()
  }

  public constructor(app: Framework) {
    this._app = () => app
  }

  @bind
  public normalizeEntry(c: Registrable): NormalizedPlugin {
    return isString(c)
      ? ([c, {}] as NormalizedPlugin)
      : (c as NormalizedPlugin)
  }

  @bind
  public setPlugin(plugin: Registrable): this {
    plugin = this.normalizeEntry(plugin)

    this.plugins = {...this.plugins, [plugin[0]]: plugin}

    return this
  }

  @bind
  public setPlugins(
    plugins: Array<NormalizedPlugin | string>,
  ): this {
    plugins.map(this.setPlugin)

    return this
  }

  @bind
  public setPreset(preset: Registrable): this {
    preset = this.normalizeEntry(preset)

    this.presets = {
      ...this.presets,
      [preset[0]]: preset,
    }

    return this
  }

  @bind
  public setPresets(
    presets: Array<NormalizedPlugin | string>,
  ): this {
    presets.map(this.setPreset)

    return this
  }

  @bind
  public unsetPreset(preset: string) {
    !this.presets[preset]
      ? this.app.error(`${preset} not found`)
      : delete this.presets[preset]

    return this
  }

  @bind
  public unsetPlugin(plugin: string) {
    !this.plugins[plugin]
      ? this.app.error(`${plugin} not found`)
      : delete this.plugins[plugin]

    return this
  }

  @bind
  public setPluginOptions(plugin: string, options: any): this {
    this.plugins[plugin] = [this.plugins[plugin][0], options]

    return this
  }

  @bind
  public setPresetOptions(preset: string, options: any): this {
    this.presets[preset] = [this.presets[preset][0], options]

    return this
  }
}
