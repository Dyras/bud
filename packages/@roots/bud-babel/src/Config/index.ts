import type {Framework} from '@roots/bud-framework'
import {boundMethod as bind} from 'autobind-decorator'
import {isString} from 'lodash'

export class Config implements Framework.Api.Babel {
  public name = '@roots/bud-babel'

  public _app: () => Framework

  public _plugins: Framework.Api.Babel.Registry = {}

  public _presets: Framework.Api.Babel.Registry = {}

  public get app() {
    return this._app()
  }

  public get plugins() {
    return this._plugins
  }

  public set plugins(plugins) {
    this._plugins = plugins
  }

  public get presets() {
    return this._presets
  }

  public set presets(presets) {
    this._presets = presets
  }

  @bind
  public init(app: Framework): this {
    this._app = () => app

    return this
  }

  @bind
  public normalizeEntry(
    c: Framework.Api.Babel.Registrable,
  ): Framework.Api.Babel.NormalizedPlugin {
    return isString(c)
      ? ([c, {}] as Framework.Api.Babel.NormalizedPlugin)
      : (c as Framework.Api.Babel.NormalizedPlugin)
  }

  @bind
  public setPlugin(
    plugin: Framework.Api.Babel.Registrable,
  ): this {
    plugin = this.normalizeEntry(plugin)

    this.plugins = {...this.plugins, [plugin[0]]: plugin}

    return this
  }

  @bind
  public setPlugins(
    plugins: Array<
      Framework.Api.Babel.NormalizedPlugin | string
    >,
  ): this {
    plugins.map(this.setPlugin)
    return this
  }

  @bind
  public setPreset(
    preset: Framework.Api.Babel.Registrable,
  ): this {
    preset = this.normalizeEntry(preset)

    this.presets = {
      ...this.presets,
      [preset[0]]: preset,
    }

    return this
  }

  @bind
  public setPresets(
    presets: Array<
      Framework.Api.Babel.NormalizedPlugin | string
    >,
  ): this {
    presets.map(this.setPreset)

    return this
  }

  @bind
  public unsetPreset(preset: string) {
    !this.presets[preset]
      ? this.app.error(`${preset} not found`)
      : delete this.presets[preset]

    return this
  }

  @bind
  public unsetPlugin(plugin: string) {
    !this.plugins[plugin]
      ? this.app.error(`${plugin} not found`)
      : delete this.plugins[plugin]

    return this
  }

  @bind
  public setPluginOptions(plugin: string, options: any): this {
    this.plugins[plugin] = [this.plugins[plugin][0], options]

    return this
  }

  @bind
  public setPresetOptions(preset: string, options: any): this {
    this.presets[preset] = [this.presets[preset][0], options]

    return this
  }
}
