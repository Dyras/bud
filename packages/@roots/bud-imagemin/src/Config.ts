/**
 * @module @roots/bud-imagemin
 */

import type {Framework} from '@roots/bud-framework'
import type {PluginOptions} from 'image-minimizer-webpack-plugin/types'

/**
 * @class Config
 */
class Config {
  public _app: () => Framework

  public get app(): Framework {
    return this._app()
  }

  public constructor(app: Framework) {
    this._app = () => app
  }

  public plugins(plugins: [string, any]) {
    this.app.hooks.on(
      'extension/image-minimizer-webpack-plugin/options',
      (options: PluginOptions) => ({
        ...(options ?? {}),
        minimizerOptions: {
          ...(options.minimizerOptions ?? {}),
          plugins,
        },
      }),
    )

    return this.app
  }
}

/**
 * @exports Config
 */
export {Config}
