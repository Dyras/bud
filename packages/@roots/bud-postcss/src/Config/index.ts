import {boundMethod as bind} from 'autobind-decorator'
import {isString} from 'lodash'
import {PluginCreator} from 'postcss'

interface Registry {
  [key: string]: [PluginCreator<any>, any]
}

interface PostCssConfig {
  /**
   * Registered plugins
   */
  plugins: Registry

  /**
   * Set a plugin
   */
  setPlugin(plugin: string | [string, any]): this

  /**
   * Set plugins
   */
  setPlugins(plugins: Array<[string, any] | string>): this

  /**
   * Set plugin options
   */
  setPluginOptions(plugin: string, options: any): this

  /**
   * Remove a plugin
   */
  unsetPlugin(plugin: string): this
}

class PostCssConfig {
  public plugins: Registry = {}

  @bind
  public normalizeEntry(
    c: string | [string, any],
  ): [string, any] {
    return isString(c) ? [c, {}] : c
  }

  @bind
  public setPlugin(plugin: string | [string, any]): this {
    const [name, config]: [string, any] =
      this.normalizeEntry(plugin)

    const pluginCreator: PluginCreator<any> = require(name)

    this.plugins = {
      ...this.plugins,
      [`${name}`]: [pluginCreator, config ?? undefined],
    }

    return this
  }

  @bind
  public setPlugins(
    plugins: Array<string | [string, any]>,
  ): this {
    plugins.map(this.setPlugin)
    return this
  }

  @bind
  public unsetPlugin(plugin: string) {
    this.plugins[plugin] && delete this.plugins[plugin]
    return this
  }

  @bind
  public setPluginOptions(plugin: string, options: any): this {
    this.plugins[plugin] = [
      this.plugins[plugin].shift(),
      options,
    ]

    return this
  }
}

export default PostCssConfig
