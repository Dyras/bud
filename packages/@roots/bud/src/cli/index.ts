/**
 * Mandatory export for @oclif/cli
 */
export {run} from '@oclif/command'
