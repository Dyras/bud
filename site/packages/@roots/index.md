---
id: 'packages'
title: '@roots/bud packages'
slug: '/'
sidebar_label: 'Packages'
sidebar_position: 0.5
custom_edit_url: null
---
