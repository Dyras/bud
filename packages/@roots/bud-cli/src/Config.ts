import {Framework} from '@roots/bud-framework'
import {Error} from '@roots/bud-dashboard'
import {CLI} from './CLI'

/**
 * Command base class
 */
export default class Config {
  public app: Framework
  public cli: CLI

  public constructor(app: Framework, cli: CLI) {
    this.app = app.bootstrap().register().boot()
    this.cli = cli
  }

  public get project() {
    return this.app.disk.get('project')
  }

  public get jsonName() {
    return `${this.cli.name}.config.json`
  }

  public get fluentName() {
    return `${this.cli.name}.config.js`
  }

  /**
   * Preflight check
   */
  public run() {
    // Guard against multi config
    this.project.has(this.jsonName) &&
      this.project.has(this.fluentName) &&
      errorMultiConf.bind(this)()

    // Guard against no config
    !this.project.has(this.jsonName) &&
      !this.project.has(this.fluentName) &&
      errorNoConf.bind(this)()

    // Run
    this.project.has(this.fluentName)
      ? require(this.project.get(this.fluentName))
      : this.app
          .entry(
            this.project.readJson(this.jsonName).entrypoints,
          )
          .run()

    return true
  }
}

function errorMultiConf() {
  Error(
    `Project contains both a ${this.jsonName} and ${this.fluentName}. They are mutually exclusive.`,
    'Multiple config sources found.',
  )
}

function errorNoConf() {
  Error(
    `
Project doesn't seem to have a config. If you need a starter config run:

$ ${this.cli.name} publish @roots/bud-support ${this.fluentName}`,
    'No config sources found.',
  )
}
