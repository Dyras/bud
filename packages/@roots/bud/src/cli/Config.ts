import {Module} from '@roots/bud-framework'
import {
  cosmiconfig,
  cosmiconfigTsLoader,
} from '@roots/bud-support'

import Bud from '../Bud'

/**
 * Configuration class
 *
 * @internal
 */
export default class Config {
  /**
   * {@link Bud} instance
   */
  public app: Bud

  /**
   * {@link cosmiconfig.Options}
   */
  public options: cosmiconfig.Options

  /**
   * Class constructor
   *
   * @param app - {@link Bud.implementation}
   * @param searchPlaces - cosmiconfig search patterns
   */
  public constructor(
    app: Bud,
    searchPlaces: cosmiconfig.Options['searchPlaces'],
  ) {
    this.app = app

    this.options = {
      searchPlaces,
      packageProp: this.app.name,
      loaders: {
        '.ts': cosmiconfigTsLoader,
      },
    }

    this.get = this.get.bind(this)
    this.apply = this.apply.bind(this)
  }

  public async get() {
    const res = await cosmiconfig
      .cosmiconfig(this.app.name, this.options)
      .search()

    return (
      res?.config ??
      function (app) {
        return app
      }
    )
  }

  public async apply() {
    const config = await this.get()

    if (!config) return this.app

    if (config.extensions) {
      config.extensions.map(mod => {
        const ext: Module = require(mod)
        this.app.use(ext as any)
      })
    }

    Object.entries(config)
      .filter(([key]) => key !== 'extensions')
      .forEach(([key, value]) => {
        this.app[key] && this.app[key](value)
      })

    return this.app
  }
}
