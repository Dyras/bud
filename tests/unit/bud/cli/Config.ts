/* eslint-disable no-console */
import {Config} from '@roots/bud/src/cli/Config'
import {Framework} from '@roots/bud-framework'

import {mocks} from './__mocks__/index'

describe('Config', () => {
  let config: Config
  let bud: Framework

  beforeEach(async () => {
    bud = mocks.bud as unknown as Framework

    config = new Config(bud, [
      process.cwd().concat('/app/babel/bud.config.js'),
    ])
  })

  afterEach(() => {
    // spies
    jest.restoreAllMocks()
    // mocked fn
    jest.clearAllMocks()
  })

  it('get', () => {
    expect(config.get).toBeInstanceOf(Function)
  })

  it('get return config', async () => {
    const conf = await config.get()
    expect(conf).toBeInstanceOf(Object)
    expect(conf).toMatchSnapshot()
  })

  it('apply', () => {
    expect(config.apply).toBeInstanceOf(Function)
  })

  it('apply returns bud', async () => {
    const returnedValue = await config.apply()
    expect(returnedValue).toMatchSnapshot(bud)
  })

  it('returns function that behaves predictably', async () => {
    const userConf = await new Config(bud, [
      `${__dirname}/__mocks__/bud.config.js`,
    ]).get()

    expect(userConf).toBeInstanceOf(Function)

    userConf(bud)

    expect(bud.tap).toHaveBeenCalled()
    expect(bud.use).toHaveBeenCalled()
    expect(bud.entry).toHaveBeenCalled()
  })
})
