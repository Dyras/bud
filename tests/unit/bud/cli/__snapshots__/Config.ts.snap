// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config apply returns bud 1`] = `
Object {
  "close": [MockFunction],
  "dashboard": Object {
    "render": [MockFunction],
  },
  "entry": [MockFunction],
  "mode": "production",
  "name": "bud",
  "project": Object {
    "getValues": [MockFunction],
    "has": [MockFunction],
    "hasPeerDependency": [MockFunction],
    "peers": Object {
      "install": [MockFunction],
    },
  },
  "tap": [MockFunction],
  "use": [MockFunction],
}
`;

exports[`Config get return config 1`] = `[Function]`;
